onmt/translate/Translator.py

```diff
@@ -23,39 +31,115 @@ class Translator(object):
        cuda (bool): use cuda
        beam_trace (bool): trace beam search for debugging
     """
-    def __init__(self, model, fields,
-                 beam_size, n_best=1,
-                 max_length=100,
-                 global_scorer=None,
-                 copy_attn=False,
-                 cuda=False,
-                 beam_trace=False,
-                 min_length=0,
-                 stepwise_penalty=False,
-                 block_ngram_repeat=0,
-                 ignore_when_blocking=[]):
+    def __init__(self, opt, report_score=True, out_file=None):
+        self.opt = opt
+        self.report_score = report_score
+
+        dummy_parser = argparse.ArgumentParser(description='train.py')
+        onmt.opts.model_opts(dummy_parser)
+        dummy_opt = dummy_parser.parse_known_args([])[0]
+
+        self.opt.cuda = opt.gpu > -1
+        if self.opt.cuda:
+            torch.cuda.set_device(self.opt.gpu)
+
+        fields, model, model_opt = \
+            onmt.ModelConstructor.load_test_model(opt, dummy_opt.__dict__)
+
+        # File to write sentences to.
+        self.out_file = out_file
+        if out_file is None:
+            self.out_file = codecs.open(opt.output, 'w', 'utf-8')
+
+        # Scorer
+        self.global_scorer = onmt.translate.GNMTGlobalScorer(
+            self.opt.alpha, self.opt.beta, self.opt.coverage_penalty,
+            self.opt.length_penalty)
+
         self.model = model
         self.fields = fields
-        self.n_best = n_best
-        self.max_length = max_length
-        self.global_scorer = global_scorer
-        self.copy_attn = copy_attn
-        self.beam_size = beam_size
-        self.cuda = cuda
-        self.min_length = min_length
-        self.stepwise_penalty = stepwise_penalty
-        self.block_ngram_repeat = block_ngram_repeat
-        self.ignore_when_blocking = set(ignore_when_blocking)
+        self.n_best = opt.n_best
+        self.max_length = opt.max_length
+        self.copy_attn = model_opt.copy_attn
+        self.beam_size = opt.beam_size
+        self.cuda = opt.cuda
+        self.min_length = opt.min_length
+        self.stepwise_penalty = opt.stepwise_penalty
+        self.beam_trace = opt.dump_beam != ""
+        self.block_ngram_repeat = opt.block_ngram_repeat
+        self.ignore_when_blocking = set(opt.ignore_when_blocking)
 
         # for debugging
         self.beam_accum = None
-        if beam_trace:
+        if self.beam_trace:
             self.beam_accum = {
                 "predicted_ids": [],
                 "beam_parent_ids": [],
                 "scores": [],
                 "log_probs": []}
 
+    def translate(self, src_dir, src_path, tgt_path):
+        data = onmt.io.build_dataset(self.fields,
```

Comment:
Translator should also now know about directories or files. Please move this out of the class as well.

Reply:
It don't get it. Moreover, I think it would be a good thing to be able to translate without passing a file name i.e. to translate from a list of sequences. Note that in this precise case of REST server we get a list, that we write on disk to provide a filepath, which does not really make sense.